File: test/notification_service/discord/formatter_test.exs
defmodule NotificationService.Discord.FormatterTest do
  use ExUnit.Case, async: true

  alias NotificationService.Discord.Formatter
  alias NotificationService.Test.Fixtures

  describe "format/2 - CI messages" do
    test "formats ci.success with green color" do
      # Arrange
      payload = Fixtures.ci_success_payload()

      # Act
      embed = Formatter.format("ci.success", payload)

      # Assert
      assert embed.title == "CI: xmer/test-repo"
      assert embed.color == 0x00FF00
      assert embed.url == "https://github.com/xmer/test-repo/actions/runs/123"
      assert length(embed.fields) == 2

      assert Enum.find(embed.fields, &(&1.name == "Status")).value == "complete"
      assert Enum.find(embed.fields, &(&1.name == "Result")).value == "success"
    end

    test "formats ci.failure with red color" do
      # Arrange
      payload = Fixtures.ci_failure_payload()

      # Act
      embed = Formatter.format("ci.failure", payload)

      # Assert
      assert embed.title == "CI: xmer/test-repo"
      assert embed.color == 0xFF0000

      assert Enum.find(embed.fields, &(&1.name == "Result")).value == "failure"
    end

    test "handles missing repository with default" do
      # Arrange
      payload = %{}

      # Act
      embed = Formatter.format("ci.success", payload)

      # Assert
      assert embed.title == "CI: unknown"
    end
  end

  describe "format/2 - PR messages" do
    test "formats pr.opened with yellow color" do
      # Arrange
      payload = Fixtures.pr_opened_payload()

      # Act
      embed = Formatter.format("pr.opened", payload)

      # Assert
      assert embed.title == "PR #42: Add new feature"
      assert embed.description == "opened by @xmer"
      assert embed.color == 0xFFFF00
      assert embed.url == "https://github.com/xmer/test-repo/pull/42"
    end

    test "formats pr.merged with purple color" do
      # Arrange
      payload = Fixtures.pr_merged_payload()

      # Act
      embed = Formatter.format("pr.merged", payload)

      # Assert
      assert embed.color == 0x800080
      assert embed.description == "merged by @xmer"
    end

    test "formats pr.closed with red color" do
      # Arrange
      payload = Fixtures.pr_closed_payload()

      # Act
      embed = Formatter.format("pr.closed", payload)

      # Assert
      assert embed.color == 0xFF0000
      assert embed.description == "closed by @xmer"
    end
  end

  describe "format/2 - download messages" do
    test "formats downloads.complete with green color and size fields" do
      # Arrange
      payload = Fixtures.download_complete_payload()

      # Act
      embed = Formatter.format("downloads.complete", payload)

      # Assert
      assert embed.title == "Download Complete: Example.Torrent"
      assert embed.color == 0x00FF00

      size_field = Enum.find(embed.fields, &(&1.name == "Size"))
      assert size_field.value == "1.0 GB"

      category_field = Enum.find(embed.fields, &(&1.name == "Category"))
      assert category_field.value == "movies"

      path_field = Enum.find(embed.fields, &(&1.name == "Path"))
      assert path_field.value == "/downloads/movies/Example.Torrent"
    end

    test "formats downloads.removed with orange color" do
      # Arrange
      payload = Fixtures.download_removed_payload()

      # Act
      embed = Formatter.format("downloads.removed", payload)

      # Assert
      assert embed.title == "Download Removed: Example.Torrent"
      assert embed.color == 0xFF8C00
    end

    test "formats size in MB" do
      # Arrange
      payload = Fixtures.download_complete_payload(%{"size" => 5_242_880})

      # Act
      embed = Formatter.format("downloads.complete", payload)

      # Assert
      size_field = Enum.find(embed.fields, &(&1.name == "Size"))
      assert size_field.value == "5.0 MB"
    end

    test "formats size in KB" do
      # Arrange
      payload = Fixtures.download_complete_payload(%{"size" => 2048})

      # Act
      embed = Formatter.format("downloads.complete", payload)

      # Assert
      size_field = Enum.find(embed.fields, &(&1.name == "Size"))
      assert size_field.value == "2.0 KB"
    end

    test "formats size in bytes" do
      # Arrange
      payload = Fixtures.download_complete_payload(%{"size" => 512})

      # Act
      embed = Formatter.format("downloads.complete", payload)

      # Assert
      size_field = Enum.find(embed.fields, &(&1.name == "Size"))
      assert size_field.value == "512 B"
    end

    test "omits nil fields" do
      # Arrange
      payload = %{"name" => "Test"}

      # Act
      embed = Formatter.format("downloads.complete", payload)

      # Assert
      assert embed.title == "Download Complete: Test"
      assert embed.fields == []
    end
  end

  describe "format/2 - deploy messages" do
    test "formats deploy.success with green color" do
      # Arrange
      payload = Fixtures.deploy_success_payload()

      # Act
      embed = Formatter.format("deploy.success", payload)

      # Assert
      assert embed.title == "Deploy: xmer/test-repo"
      assert embed.color == 0x00FF00

      status_field = Enum.find(embed.fields, &(&1.name == "Status"))
      assert status_field.value == "success"

      duration_field = Enum.find(embed.fields, &(&1.name == "Duration"))
      assert duration_field.value == "45s"
    end

    test "formats deploy.failure with red color" do
      # Arrange
      payload = Fixtures.deploy_failure_payload()

      # Act
      embed = Formatter.format("deploy.failure", payload)

      # Assert
      assert embed.title == "Deploy: xmer/test-repo"
      assert embed.color == 0xFF0000
    end

    test "omits duration field when not present" do
      # Arrange
      payload = %{"repository" => "xmer/test-repo"}

      # Act
      embed = Formatter.format("deploy.success", payload)

      # Assert
      refute Enum.any?(embed.fields, &(&1.name == "Duration"))
    end
  end

  describe "format/2 - DLQ alerts" do
    test "formats DLQ alert with service name and retry count" do
      # Arrange
      payload = Fixtures.dlq_alert_payload()

      # Act
      embed = Formatter.format("dlq.qbittorrent", payload)

      # Assert
      assert embed.title == "DLQ Alert: qbittorrent-consumer"
      assert embed.description == "downloads.add queue exceeded retry limit"
      assert embed.color == 0xFF0000

      retries_field = Enum.find(embed.fields, &(&1.name == "Retries"))
      assert retries_field.value == "20"

      queue_field = Enum.find(embed.fields, &(&1.name == "Queue"))
      assert queue_field.value == "downloads.add"

      error_field = Enum.find(embed.fields, &(&1.name == "Error"))
      assert error_field.value == "qBittorrent API timeout"
    end
  end

  describe "format/2 - polling failure" do
    test "formats polling failure with error details" do
      # Arrange
      payload = Fixtures.polling_failure_payload()

      # Act
      embed = Formatter.format("polling.failure", payload)

      # Assert
      assert embed.title == "Polling Failure: qbittorrent-consumer"
      assert embed.color == 0xFF0000

      error_field = Enum.find(embed.fields, &(&1.name == "Error"))
      assert error_field.value == "qBittorrent API unreachable for 10+ minutes"
    end
  end

  describe "format/2 - unknown messages" do
    test "formats unknown routing key with raw JSON payload" do
      # Arrange
      payload = %{"foo" => "bar"}

      # Act
      embed = Formatter.format("some.unknown.key", payload)

      # Assert
      assert embed.title == "Notification"
      assert embed.color == 0x0099FF
      assert String.contains?(embed.description, "foo")
    end

    test "formats unknown with non-serializable payload using inspect" do
      # Arrange - create a payload that Jason can't encode
      # A function reference cannot be encoded
      payload = %{func: &String.length/1}

      # Act
      embed = Formatter.format("some.unknown.key", payload)

      # Assert
      assert embed.title == "Notification"
      assert is_binary(embed.description)
    end
  end

  describe "format_bytes edge cases" do
    test "handles non-numeric size gracefully" do
      # Arrange
      payload = %{"name" => "Test", "size" => "not a number"}

      # Act
      embed = Formatter.format("downloads.complete", payload)

      # Assert
      size_field = Enum.find(embed.fields, &(&1.name == "Size"))
      assert size_field.value == "Unknown"
    end
  end
end

File: tests/discord/formatter.test.ts
import { describe, expect, it } from "bun:test";
import { formatEmbed } from "../../src/discord/formatter.js";

describe("formatEmbed", () => {
	describe("CI messages", () => {
		it("should format ci.success message", () => {
			// Arrange
			const payload = {
				repository: "owner/repo",
				status: "complete",
				run_url: "https://github.com/owner/repo/actions/runs/123",
			};

			// Act
			const embed = formatEmbed("ci.success", payload);

			// Assert
			expect(embed.title).toBe("CI Build Succeeded");
			expect(embed.color).toBe(0x57f287);
			expect(embed.url).toBe(payload.run_url);
			expect(embed.fields).toContainEqual({
				name: "Repository",
				value: "owner/repo",
				inline: true,
			});
			expect(embed.fields).toContainEqual({ name: "Result", value: "success", inline: true });
		});

		it("should format ci.failure message", () => {
			// Arrange
			const payload = {
				repository: "owner/repo",
				status: "complete",
				run_url: "https://github.com/owner/repo/actions/runs/123",
			};

			// Act
			const embed = formatEmbed("ci.failure", payload);

			// Assert
			expect(embed.title).toBe("CI Build Failed");
			expect(embed.color).toBe(0xed4245);
			expect(embed.fields).toContainEqual({ name: "Result", value: "failure", inline: true });
		});
	});

	describe("PR messages", () => {
		it("should format pr.opened message", () => {
			// Arrange
			const payload = {
				pr_number: 42,
				pr_title: "Add new feature",
				author: "contributor",
				pr_url: "https://github.com/owner/repo/pull/42",
			};

			// Act
			const embed = formatEmbed("pr.opened", payload);

			// Assert
			expect(embed.title).toBe("Pull Request Opened");
			expect(embed.color).toBe(0xfee75c);
			expect(embed.url).toBe(payload.pr_url);
			expect(embed.fields).toContainEqual({ name: "PR", value: "#42", inline: true });
			expect(embed.fields).toContainEqual({ name: "Author", value: "contributor", inline: true });
		});

		it("should format pr.merged message", () => {
			// Arrange
			const payload = {
				pr_number: 42,
				pr_title: "Add new feature",
				author: "contributor",
			};

			// Act
			const embed = formatEmbed("pr.merged", payload);

			// Assert
			expect(embed.title).toBe("Pull Request Merged");
			expect(embed.color).toBe(0x9b59b6);
		});

		it("should format pr.closed message", () => {
			// Arrange
			const payload = {
				pr_number: 42,
				pr_title: "Add new feature",
				author: "contributor",
			};

			// Act
			const embed = formatEmbed("pr.closed", payload);

			// Assert
			expect(embed.title).toBe("Pull Request Closed");
			expect(embed.color).toBe(0xed4245);
		});

		it("should truncate long PR titles", () => {
			// Arrange
			const longTitle = "A".repeat(150);
			const payload = { pr_number: 1, pr_title: longTitle, author: "user" };

			// Act
			const embed = formatEmbed("pr.opened", payload);

			// Assert
			const titleField = embed.fields?.find((f) => f.name === "Title");
			expect(titleField?.value.length).toBeLessThanOrEqual(100);
			expect(titleField?.value).toEndWith("...");
		});
	});

	describe("Download messages", () => {
		it("should format downloads.complete message", () => {
			// Arrange
			const payload = {
				name: "Media File",
				size: 5368709120,
				category: "TV",
				savePath: "/downloads/tv",
			};

			// Act
			const embed = formatEmbed("downloads.complete", payload);

			// Assert
			expect(embed.title).toBe("Download Complete");
			expect(embed.color).toBe(0x57f287);
			expect(embed.fields).toContainEqual({ name: "Name", value: "Media File", inline: false });
			expect(embed.fields).toContainEqual({ name: "Size", value: "5 GB", inline: true });
			expect(embed.fields).toContainEqual({ name: "Category", value: "TV", inline: true });
		});

		it("should format downloads.removed message", () => {
			// Arrange
			const payload = { name: "Old File", size: 1024, category: "Movies" };

			// Act
			const embed = formatEmbed("downloads.removed", payload);

			// Assert
			expect(embed.title).toBe("Download Removed");
			expect(embed.color).toBe(0xe67e22);
		});
	});

	describe("Deploy messages", () => {
		it("should format deploy.success message", () => {
			// Arrange
			const payload = { repository: "owner/repo", status: "complete", duration: 45 };

			// Act
			const embed = formatEmbed("deploy.success", payload);

			// Assert
			expect(embed.title).toBe("Deployment Succeeded");
			expect(embed.color).toBe(0x57f287);
			expect(embed.fields).toContainEqual({ name: "Duration", value: "45s", inline: true });
		});

		it("should format deploy.failure message", () => {
			// Arrange
			const payload = { repository: "owner/repo", status: "failed" };

			// Act
			const embed = formatEmbed("deploy.failure", payload);

			// Assert
			expect(embed.title).toBe("Deployment Failed");
			expect(embed.color).toBe(0xed4245);
		});
	});

	describe("Polling messages", () => {
		it("should format polling.failure message", () => {
			// Arrange
			const payload = { service: "api-service", error: "Connection timeout" };

			// Act
			const embed = formatEmbed("polling.failure", payload);

			// Assert
			expect(embed.title).toBe("Service Polling Failed");
			expect(embed.color).toBe(0xed4245);
			expect(embed.fields).toContainEqual({ name: "Service", value: "api-service", inline: true });
		});
	});

	describe("DLQ messages", () => {
		it("should format dlq.* message", () => {
			// Arrange
			const payload = {
				service: "my-service",
				queue: "my-queue",
				error: "Processing failed",
				retryCount: 20,
			};

			// Act
			const embed = formatEmbed("dlq.my-service", payload);

			// Assert
			expect(embed.title).toBe("Dead Letter Queue Alert");
			expect(embed.color).toBe(0xed4245);
			expect(embed.fields).toContainEqual({ name: "Service", value: "my-service", inline: true });
			expect(embed.fields).toContainEqual({ name: "Retry Count", value: "20", inline: true });
		});

		it("should format notifications.dlq.* message", () => {
			// Arrange
			const payload = {
				service: "notification-service",
				queue: "notifications",
				error: "Discord error",
				retryCount: 5,
			};

			// Act
			const embed = formatEmbed("notifications.dlq.notification-service", payload);

			// Assert
			expect(embed.title).toBe("Dead Letter Queue Alert");
		});
	});

	describe("Disk space messages", () => {
		it("should format warn.diskspace message", () => {
			// Arrange
			const payload = {
				volume: "/share/Downloads",
				mount_point: "/share/Downloads",
				used_percent: 82,
				free_bytes: 90000000000,
				threshold: "warning",
			};

			// Act
			const embed = formatEmbed("warn.diskspace", payload);

			// Assert
			expect(embed.title).toBe("Disk Space Warning");
			expect(embed.color).toBe(0xfee75c);
			expect(embed.fields).toContainEqual({
				name: "Volume",
				value: "/share/Downloads",
				inline: true,
			});
			expect(embed.fields).toContainEqual({ name: "Usage", value: "82%", inline: true });
		});

		it("should format error.diskspace message", () => {
			// Arrange
			const payload = {
				volume: "/share/Media",
				used_percent: 91,
				free_bytes: 45000000000,
			};

			// Act
			const embed = formatEmbed("error.diskspace", payload);

			// Assert
			expect(embed.title).toBe("Disk Space Critical");
			expect(embed.color).toBe(0xe67e22);
			expect(embed.fields).toContainEqual({ name: "Volume", value: "/share/Media", inline: true });
			expect(embed.fields).toContainEqual({ name: "Usage", value: "91%", inline: true });
		});

		it("should format critical.diskspace message", () => {
			// Arrange
			const payload = {
				volume: "/share/System",
				used_percent: 97,
				free_bytes: 15000000000,
			};

			// Act
			const embed = formatEmbed("critical.diskspace", payload);

			// Assert
			expect(embed.title).toBe("Disk Space Emergency");
			expect(embed.color).toBe(0xed4245);
			expect(embed.fields).toContainEqual({ name: "Volume", value: "/share/System", inline: true });
			expect(embed.fields).toContainEqual({ name: "Usage", value: "97%", inline: true });
		});

		it("should format info.diskspace recovery message", () => {
			// Arrange
			const payload = {
				volume: "/share/Downloads",
				used_percent: 75,
				previous_threshold: "warning",
			};

			// Act
			const embed = formatEmbed("info.diskspace", payload);

			// Assert
			expect(embed.title).toBe("Disk Space Recovery");
			expect(embed.color).toBe(0x57f287);
			expect(embed.fields).toContainEqual({
				name: "Recovered From",
				value: "warning",
				inline: true,
			});
		});

		it("should format info.diskspace status report", () => {
			// Arrange
			const payload = {
				volumes: [
					{
						volume: "/share/Downloads",
						mount_point: "/share/Downloads",
						used_percent: 45,
						status: "ok",
					},
					{
						volume: "/share/Media",
						mount_point: "/share/Media",
						used_percent: 82,
						status: "warning",
					},
				],
			};

			// Act
			const embed = formatEmbed("info.diskspace", payload);

			// Assert
			expect(embed.title).toBe("Disk Space Status");
			expect(embed.color).toBe(0x3498db);
			expect(embed.description).toContain("/share/Downloads: 45% (ok)");
			expect(embed.description).toContain("/share/Media: 82% (warning)");
		});

		it("should handle missing fields gracefully for disk space", () => {
			// Arrange
			const payload = {};

			// Act
			const embed = formatEmbed("warn.diskspace", payload);

			// Assert
			expect(embed.fields).toContainEqual({ name: "Volume", value: "unknown", inline: true });
			expect(embed.fields).toContainEqual({ name: "Usage", value: "0%", inline: true });
		});
	});

	describe("Unknown messages", () => {
		it("should format unknown message type with JSON", () => {
			// Arrange
			const payload = { customField: "value", count: 42 };

			// Act
			const embed = formatEmbed("custom.event", payload);

			// Assert
			expect(embed.title).toBe("Notification: custom.event");
			expect(embed.color).toBe(0x3498db);
			expect(embed.description).toContain("customField");
		});
	});

	describe("Missing fields", () => {
		it("should handle missing fields gracefully", () => {
			// Arrange
			const payload = {};

			// Act
			const embed = formatEmbed("ci.success", payload);

			// Assert
			expect(embed.fields).toContainEqual({ name: "Repository", value: "unknown", inline: true });
		});
	});
});
